completions/ferrishot.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH ferrishot 1  "ferrishot 0.2.0" 
.SH NAME
ferrishot \- A powerful screenshot app
.SH SYNOPSIS
\fBferrishot\fR [\fB\-r\fR|\fB\-\-region\fR] [\fB\-l\fR|\fB\-\-last\-region\fR] [\fB\-a\fR|\fB\-\-accept\-on\-select\fR] [\fB\-d\fR|\fB\-\-delay\fR] [\fB\-s\fR|\fB\-\-save\-path\fR] [\fB\-D\fR|\fB\-\-dump\-default\-config\fR] [\fB\-C\fR|\fB\-\-config\-file\fR] [\fB\-S\fR|\fB\-\-silent\fR] [\fB\-j\fR|\fB\-\-json\fR] [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fIFILE\fR] 
.SH DESCRIPTION
A powerful screenshot app
.SH OPTIONS
.TP
\fB\-r\fR, \fB\-\-region\fR=\fIWxH+X+Y\fR
Open with a region pre\-selected

Format: `<width>x<height>+<top\-left\-x>+<top\-left\-y>`

Each value can be absolute.
\- 550 for `x` means top\-left corner starts after 550px
\- 100 for `height` means it will be 100px tall

Each can also be relative to the height (for `y` and `height`) or width (for `width` and `x`)
\- 0.2 for `width` means it region takes up 20% of the width of the image.
\- 0.5 for `y` means the top\-left corner will be at the vertical center

The format can also end with 1 or 2 percentages, which shifts the region relative to the region\*(Aqs size
\- If `width` is `250`, end region with `+30%` to move right by 75px or `\-40%` to move left by 100px
\- Supplying 2 percentage at the end like `+30%\-10%`, the 1st affects x\-offset and the 2nd affects y\-offset

With the above syntax, you can create all the regions you want.
\- `100x1.0+0.5+0\-50%`: Create a 100px wide, full height, horizontally centered region
\- `1.0x1.0+0+0`: Create a region that spans the full screen. You can use alias `full` for this
.TP
\fB\-l\fR, \fB\-\-last\-region\fR
Use last region
.TP
\fB\-a\fR, \fB\-\-accept\-on\-select\fR=\fIACTION\fR
Accept capture and perform the action as soon as a selection is made

If holding `ctrl` while you are releasing the left mouse button on the first selection,
the behavior is cancelled

It\*(Aqs quite useful to run ferrishot, select a region and have it instantly be copied to the
clipboard for example. In 90% of situations you won\*(Aqt want to do much post\-processing of
the region and this makes that experience twice as fast. You can always opt\-out with `ctrl`

Using this option with `\-\-region` or `\-\-last\-region` will run ferrishot in \*(Aqheadless mode\*(Aq,
without making a new window.
.br

.br
\fIPossible values:\fR
.RS 14
.IP \(bu 2
copy: Copy image to the clipboard
.IP \(bu 2
save: Save image to a file
.IP \(bu 2
upload: Upload image to the internet
.RE
.TP
\fB\-d\fR, \fB\-\-delay\fR=\fIMILLISECONDS\fR
Wait this long before launch
.TP
\fB\-s\fR, \fB\-\-save\-path\fR=\fIPATH\fR
Instead of opening a file picker to save the screenshot, save it to this path instead
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help (see a summary with \*(Aq\-h\*(Aq)
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.SH CONFIG
.TP
\fB\-D\fR, \fB\-\-dump\-default\-config\fR
Write contents of the default config to /home/e/.config/ferrishot.kdl
.TP
\fB\-C\fR, \fB\-\-config\-file\fR=\fIFILE.KDL\fR [default: /home/e/.config/ferrishot.kdl]
Use the provided config file
.SH OUTPUT
.TP
\fB\-S\fR, \fB\-\-silent\fR
Run in silent mode. Do not print anything
.TP
\fB\-j\fR, \fB\-\-json\fR
Print in JSON format
.SH VERSION
v0.2.0
.SH AUTHORS
Nik Revenco


CHANGELOG.md
# v0.3.0 - 4 May 2025

- Added support for touch inputs
- Ferrishot can now be configured! The config format uses [KDL](https://kdl.dev/). Here is an excerpt from the default config:

```kdl
// Show the size indicator
size-indicator #true
// Show icons around the selection
selection-icons #true

keys {
  // Leave the app
  exit key=<esc>

  // Copies selected region to clipboard, exiting
  copy-to-clipboard mod=ctrl key=c
  copy-to-clipboard key=<enter>
}

theme {
  // color of the frame around the selection
  selection-frame 0xab_61_37

  // background color of the region that is not selected
  non-selected-region 0x00_00_00 opacity=0.5
}
```

- Image upload. Press `ctrl + u` to upload your screenshot to the internet. You'll get a preview of the image, link to it and a QR Code.
- Vim keybindings! Ferrishot is fully controllable via keyboard, if you want to do that
  - `h | j | k | l` to move by 1 pixel the given side
  - `H | J | K | L` to extend by 1 pixel the given side
  - `ctrl + h | j | k | l` shrinks by 1 pixel the given side
  - Use `alt` in combination with any of the above keys to do the movement of `125px` instead of `1px`
  - There is also `gg` to move to top-left corner, `G` for bottom-right corner and more. You can view a cheatsheet in-app by pressing `?`.
  - All the keybindings can be overridden with a custom config. You are also able to define custom ones, if you want it to be `50px` instead of `125px` for all of them simply do `ferrishot --dump-default-config` and then edit the values `125 -> 50`.
- Super-select mode, press `t` and:
  - The screen will be divided into 25 regions each assigned a letter. Press any of the letters, then:
  - The chosen region will be further divided into 25 more regions. This will repeat yet again for a 3rd time.
  - Allows you to place your cursor into 1 of 15,625 positions on the screen within 4 keystrokes. This will select the top-left corner
  - Repeat this with the bottom-right corner by using `b`. Now you have selected your screenshot without using a mouse, in just 8 keystrokes!
- More powerful command line interface

  - `--delay` wait some time before taking a screenshot
  - `--region` open program with a custom region
  - `--save-path` choose a file to save the image to, instead of opening file picker (when using `ctrl + s`)
  - `--accept-on-select` instantly save, upload or copy image to clipboard as soon as you create the first selection.

    This can be cancelled with `ctrl`. It's handy since often you want to instantly copy the selection to clipboard and not do anything fancy.

    So you can therefore set a shell alias like `alias ferrishot=ferrishot --accept-on-select=copy` and that may be satisfying to you 90% of the time.

    In some cases you'd like to save the image, so just hold ctrl to disable this behaviour when releasing the left-mouse button.

    Using this option with `--region` will run ferrishot in "headless mode", without opening a window (as the region created with `--region` _will_ be the first region)

- Ferrishot now has a website! [ferrishot.com](https://ferrishot.com)
- The command-line interface has completions for several shells. Completions for each shell can be found in `completions/` folder on GitHub:
  - PowerShell
  - Bash
  - Elv
  - Fish
  - Nushell

##### Breaking

Removed `--instant` flag, use instead:

```
ferrishot --accept-on-select copy
```

# v0.2.0 - 16 April 2025

- Right-click will snap the closest corner to the current cursor position
- There are now buttons available for these actions:
  - `Enter`: Copy to Clipboard
  - `F11`: Select entire monitor
  - `Ctrl + S`: Save Screenshot
  - `Esc`: Exit
- Added an indicator for the width and height of selection in the bottom right corner. This indicator can be edited to set the selection to a concrete size!
- Holding `Shift` while resizing or moving the selection will now do it 10 times slower, to allow being very accurate.
- Improved security on linux as per review from a reddit user: <https://www.reddit.com/r/rust/comments/1jxwd86/comment/mmvfzhy/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button>

# v0.1.0 - 12 April 2025

The initial release comes with the following features:

- Select a region on the screen by left clicking and drag
- Resize region by dragging on any of the sides
- Move the region around by dragging in the center
- `Esc` closes the app
- `Enter` or `Ctrl c` copy region to clipboard
- `Ctrl s` save region as an image to path
- Instantly copy region to clipboard with `--instant` flag


completions/ferrishot.ts
const completion: Fig.Spec = {
  name: "ferrishot",
  description: "A powerful screenshot app",
  options: [
    {
      name: ["-r", "--region"],
      description: "Open with a region pre-selected",
      isRepeatable: true,
      args: {
        name: "region",
        isOptional: true,
      },
    },
    {
      name: ["-a", "--accept-on-select"],
      description: "Accept on first selection",
      isRepeatable: true,
      args: {
        name: "accept_on_select",
        isOptional: true,
        suggestions: [
          {
            name: "copy",
            description: "Copy image to the clipboard",
          },
          {
            name: "save",
            description: "Save image to a file",
          },
          {
            name: "upload",
            description: "Upload image to the internet",
          },
        ],
      },
    },
    {
      name: ["-d", "--delay"],
      description: "Wait this long before launch",
      isRepeatable: true,
      args: {
        name: "delay",
        isOptional: true,
      },
    },
    {
      name: ["-s", "--save-path"],
      description: "Save image to path",
      isRepeatable: true,
      args: {
        name: "save_path",
        isOptional: true,
        template: "filepaths",
      },
    },
    {
      name: ["-C", "--config-file"],
      description: "Use the provided config file",
      isRepeatable: true,
      args: {
        name: "config_file",
        isOptional: true,
        template: "filepaths",
      },
    },
    {
      name: "--log-level",
      description: "Choose a miniumum level at which to log",
      hidden: true,
      isRepeatable: true,
      args: {
        name: "log_level",
        isOptional: true,
      },
    },
    {
      name: "--log-file",
      description: "Path to the log file",
      hidden: true,
      isRepeatable: true,
      args: {
        name: "log_file",
        isOptional: true,
        template: "filepaths",
      },
    },
    {
      name: "--log-filter",
      description: "Filter for specific Rust module or crate, instead of showing logs from all crates",
      hidden: true,
      isRepeatable: true,
      args: {
        name: "log_filter",
        isOptional: true,
      },
    },
    {
      name: ["-l", "--last-region"],
      description: "Use last region",
      exclusiveOn: [
        "-r",
        "--region",
      ],
    },
    {
      name: ["-D", "--dump-default-config"],
      description: "Write the default config to /home/e/.config/ferrishot.kdl",
    },
    {
      name: ["-S", "--silent"],
      description: "Run in silent mode",
    },
    {
      name: ["-j", "--json"],
      description: "Print in JSON format",
      exclusiveOn: [
        "-S",
        "--silent",
      ],
    },
    {
      name: "--log-stderr",
      description: "Log to standard error instead of file",
      exclusiveOn: [
        "-S",
        "--silent",
      ],
    },
    {
      name: "--debug",
      description: "Launch in debug mode (F12)",
    },
    {
      name: ["-h", "--help"],
      description: "Print help (see more with '--help')",
    },
    {
      name: ["-V", "--version"],
      description: "Print version",
    },
  ],
  args: {
    name: "file",
    isOptional: true,
    template: "filepaths",
  },
};

export default completion;
